File: webapp/src/types/Funnel.ts
export interface Funnel {
  id: string;
  name: string;
  description: string;
  steps: FunnelStep[];
}

export interface FunnelStep {
  id: string;
  name: string;
  step_order: number;
  definition: string;
  example: string;
  action: string;
}

export type NewFunnelStep = Omit<FunnelStep, 'id'>;
export interface NewFunnel {
  name: string;
  description: string;
  steps: NewFunnelStep[];
}
export interface ObjectFunnel {
  funnelId: any;
  stepId: string;
}

export interface FunnelUpdate {
  id: any;
  name: string;
  description: string;
  steps_create: FunnelStep[];
  steps_update: FunnelStep[];
  steps_delete: string[];
  step_mapping: { [oldId: string]: string };
}

File: server/internal/models/funnel.go
// funnel.go
package models

import (
	"database/sql"
	"time"
)

type Funnel struct {
	ID          string       `json:"id"`
	Name        string       `json:"name"`
	Description string       `json:"description"`
	CreatorID   string       `json:"creator_id"`
	OrgID       string       `json:"org_id"`
	CreatedAt   time.Time    `json:"created_at"`
	DeletedAt   sql.NullTime `json:"deleted_at"`
	Steps       []Step       `json:"steps"`
	ObjectCount int          `json:"object_count"`
}

type Step struct {
	ID          string       `json:"id"`
	FunnelID    string       `json:"funnel_id"`
	Name        string       `json:"name"`
	Definition  string       `json:"definition"`
	Example     string       `json:"example"`
	Action      string       `json:"action"`
	StepOrder   int          `json:"step_order"`
	CreatedAt   time.Time    `json:"created_at"`
	LastUpdated time.Time    `json:"last_updated"`
	DeletedAt   sql.NullTime `json:"deleted_at"`
	ObjectCount int          `json:"object_count"`
}

type FunnelUpdate struct {
	ID          string            `json:"id"`
	Name        string            `json:"name"`
	Description string            `json:"description"`
	StepsCreate []Step            `json:"steps_create"`
	StepsUpdate []Step            `json:"steps_update"`
	StepsDelete []string          `json:"steps_delete"`
	StepMapping map[string]string `json:"step_mapping"`
}